queue/claimresolver.js

```diff
@@ -0,0 +1,243 @@
+ * has been exceeded. The messages advice that if a task with the given
+ * `takenUntil` and `taskId` exists, then `runId` maybe need to be resolved by
+ * `claim-expired` and the task retried (depending on `retriesLeft`).
+ *
+ * Notice, that the task may not exists, or may not have a the given run, or
+ * it may be the case that the `takenUntil` doesn't match. These should be
+ * ignored as it implies a failed request (which wasn't retried) or a run
+ * that was reclaimed.
+ */
+class ClaimResolver {
+  /**
+   * Create ClaimResolver instance.
+   *
+   * options:
+   * {
+   *   Task:           // instance of data.Task
+   *   queueService:   // instance of QueueService
+   *   publisher:      // publisher from base.Exchanges
+   *   pollingDelay:   // Number of ms to sleep between polling
+   *   parallelism:    // Number of polling loops to run in parallel
+   *                   // Each handles up to 32 messages in parallel
+   * }
+   */
+  constructor(options) {
+    assert(options, "options must be given");
+    assert(options.Task.prototype instanceof data.Task,
+           "Expected data.Task instance");
+    assert(options.queueService instanceof QueueService,
+           "Expected instance of QueueService");
+    assert(options.publisher, "Expected a publisher");
+    assert(typeof(options.pollingDelay) === 'number',
+           "Expected pollingDelay to be a number");
+    assert(typeof(options.parallelism) === 'number',
+           "Expected parallelism to be a number");
+    this.Task         = options.Task;
+    this.queueService = options.queueService;
+    this.publisher    = options.publisher;
+    this.pollingDelay = options.pollingDelay;
+    this.parallelism  = options.parallelism;
+
+    // Promise that polling is done
+    this.done         = null;
+    // Boolean that polling should stop
+    this.stopping     = false;
+  }
+
+  /** Start polling */
+  start() {
+    if (this.done) {
+      return;
+    }
+    this.stopping = false;
+
+    // Start a loop for the amount of parallelism desired
+    var loops = [];
+    for(var i = 0; i < this.parallelism; i++) {
+      loops.push(this.poll());
+    }
+    // Create promise that we're done looping
+    this.done = Promise.all(loops).catch((err) => {
+      debug("Error: %s, as JSON: %j", err, err, err.stack);
+      throw err;
+    }).then(() => {
+      this.done = null;
+    });
+  }
+
+  /** Terminate iteration, returns promise that polling is stopped */
+  terminate() {
+    this.stopping = true;
+    return this.done;
+  }
+
+  /** Poll for messages and handle them in a loop */
+  async poll() {
+    while(!this.stopping) {
+      var messages = await this.queueService.pollClaimQueue();
+      debug("Fetched %s messages", messages.length);
+
+      await Promise.all(messages.map((message) => {
+        // Don't let a single task error break the loop, it'll be retried later
+        // as we don't remove message unless they are handled
```

Comment:
do we have facilities for tracking which tasks have failed, disappeared from the API and are never seen again?  How do those things get resolved?